File: backend/app/schemas/commit_analysis.py
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class CommitChangeType(StrEnum):
    FEATURE = "Feature"
    BUGFIX = "Bugfix"
    REFACTOR = "Refactor"
    COMMON = "Common"
    TEST = "Test"
    CI = "CI"
    OTHER = "Other"


class CommitAnalysisStatus(StrEnum):
    NOT_ANALYZED = "未分析"
    ANALYZED = "已分析"
    CLOSED = "已闭环"


class CommitAnalysisUpdate(BaseModel):
    what_commit_did: str | None = None
    change_type: CommitChangeType | None = None
    affects_api: bool | None = None
    vllm_ascend_impact: str | None = None
    next_plan: str | None = None
    planned_closure_time: str | None = None
    actual_closure_time: str | None = None


class CommitAnalysisAssignRequest(BaseModel):
    assignee: str = Field(..., max_length=100)


class CommitAnalysisBatchRequest(BaseModel):
    shas: list[str] = Field(default_factory=list, max_length=500)


class CommitAnalysisResponse(BaseModel):
    project: str
    sha: str
    assignee: str | None = None
    what_commit_did: str | None = None
    change_type: CommitChangeType | None = None
    affects_api: bool | None = None
    vllm_ascend_impact: str | None = None
    next_plan: str | None = None
    planned_closure_time: str | None = None
    actual_closure_time: str | None = None
    created_at: str | None = None
    created_by: str | None = None
    updated_at: str | None = None
    updated_by: str | None = None
    status: CommitAnalysisStatus
    can_edit: bool


class CommitAnalysisBatchItem(BaseModel):
    sha: str
    assignee: str | None = None
    change_type: CommitChangeType | None = None
    status: CommitAnalysisStatus


class CommitAnalysisBatchResponse(BaseModel):
    project: str
    analyses: dict[str, CommitAnalysisBatchItem]
    filters: dict[str, list[Any]]

File: frontend/src/services/commitAnalysis.ts
import apiClient from './api'

export const CHANGE_TYPES = ['Feature', 'Bugfix', 'Refactor', 'Common', 'Test', 'CI', 'Other'] as const
export const ANALYSIS_STATUSES = ['未分析', '已分析', '已闭环'] as const

export type CommitChangeType = typeof CHANGE_TYPES[number]
export type CommitAnalysisStatus = typeof ANALYSIS_STATUSES[number]

export interface CommitAnalysis {
  project: string
  sha: string
  assignee: string | null
  what_commit_did: string | null
  change_type: CommitChangeType | null
  affects_api: boolean | null
  vllm_ascend_impact: string | null
  next_plan: string | null
  planned_closure_time: string | null
  actual_closure_time: string | null
  created_at: string | null
  created_by: string | null
  updated_at: string | null
  updated_by: string | null
  status: CommitAnalysisStatus
  can_edit: boolean
}

export interface CommitAnalysisUpdate {
  what_commit_did?: string | null
  change_type?: CommitChangeType | null
  affects_api?: boolean | null
  vllm_ascend_impact?: string | null
  next_plan?: string | null
  planned_closure_time?: string | null
  actual_closure_time?: string | null
}

export interface CommitAnalysisBatchItem {
  sha: string
  assignee: string | null
  change_type: CommitChangeType | null
  status: CommitAnalysisStatus
}

export interface CommitAnalysisBatchResponse {
  project: string
  analyses: Record<string, CommitAnalysisBatchItem>
  filters: {
    assignees: string[]
    change_types: CommitChangeType[]
    statuses: CommitAnalysisStatus[]
  }
}

export const getCommitAnalysis = async (project: string, sha: string): Promise<CommitAnalysis> => {
  const response = await apiClient.get(`/commit-analysis/${project}/${sha}`)
  return response.data
}

export const batchGetCommitAnalysis = async (project: string, shas: string[]): Promise<CommitAnalysisBatchResponse> => {
  const response = await apiClient.post(`/commit-analysis/${project}/batch`, { shas })
  return response.data
}

export const claimCommitAnalysis = async (project: string, sha: string): Promise<CommitAnalysis> => {
  const response = await apiClient.post(`/commit-analysis/${project}/${sha}/claim`)
  return response.data
}

export const assignCommitAnalysis = async (project: string, sha: string, assignee: string): Promise<CommitAnalysis> => {
  const response = await apiClient.post(`/commit-analysis/${project}/${sha}/assign`, { assignee })
  return response.data
}

export const updateCommitAnalysis = async (
  project: string,
  sha: string,
  data: CommitAnalysisUpdate
): Promise<CommitAnalysis> => {
  const response = await apiClient.put(`/commit-analysis/${project}/${sha}`, data)
  return response.data
}
